src/pghstore/_speedups.c

```diff
@@ -0,0 +1,387 @@
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#include <stdio.h>
+#include <string.h>
+
+struct module_state {
+    PyObject *error;
+};
+
+#if PY_MAJOR_VERSION >= 3
+#define GETSTATE(m) ((struct module_state*)PyModule_GetState(m))
+#endif
+
+#define CHAR_CHECK(s, pos, c)                        \
+  if (s[pos] != char) {                              \
+    PyErr_Format(PyExc_ValueError,                   \
+                 "Encountered %c at %i, was expecting %c.",   \
+                 s[pos], pos, c);                             \
+    return NULL;                                              \
+  }
+
+#define isEven(a) (((a) & 1) == 0)
+
+#define ConcatOrGotoCleanup(result, string) { PyBytes_Concat(&result, string); if (!result) { goto _speedup_dumps_cleanup_and_exit; } }
+
+/*
+ * Just like strchr, but find first -unescaped- occurrence of c in s.
+ */
+char *
+strchr_unescaped(char *s, char c) 
+{
+  char *p = strchr(s, c), *q;
+  while (p != NULL) { /* loop through all the c's */
+    q = p; /* scan backwards through preceding escapes */
+    while (q > s && *(q-1) == '\\')
+      q--;
+    if (isEven(p-q)) /* even number of esc's => c is good */
+        return p;
+    p = strchr(p+1, c); /* else odd escapes => c is escaped, keep going */
+  }
+  return NULL;
+}
+
+/* Written to match pghstore._native.unescape but slightly smarter. */
+PyObject *
+unescape(char *copy_from_start, char *copy_from_end, char *encoding, const char *errors)
+{
+  /* NOTE(sigmavirus24): In the event that there are \s in the string, we're
+   * over-allocating here but that is okay since we're going to free shortly
+   * after use.
+   * The +1 is also necessary to ensure it's '\0' terminated.
+   */
+  ssize_t copy_max_length = (copy_from_end - copy_from_start) + 1;
+  ssize_t copy_index = 0;
+  ssize_t index;
+  char *copy_to = calloc(copy_max_length, sizeof(char));
+  PyObject *unicode;
+
+  for (index = 0; index < (copy_max_length - 1); index++) {
+      if (copy_from_start[index] == '\\' && copy_from_start[index + 1] != '\\') {
+          continue;
+      }
+      copy_to[copy_index++] = copy_from_start[index];
+  }
+  unicode = PyUnicode_Decode(copy_to, copy_index, encoding, errors);
+  free(copy_to);
+
+  return unicode;
+}
+
+static PyObject *
+_speedups_loads(PyObject *self, PyObject *args, PyObject *keywds)
+{
+  static char *keyword_argument_names[] = {"string", "encoding", "return_type", NULL};
+  const char *errors = "strict";
+  char *encoding = "utf-8";
+  char *s;
+  Py_ssize_t s_length = 0;
+  int i, null_value = 0;
+  int is_dictionary = 0;
+  char *key_start, *key_end, *value_start, *value_end;
+  PyTypeObject *return_type = &PyDict_Type;
+  PyObject *return_value, *key, *value;
+
+  if (!PyArg_ParseTupleAndKeywords(args, keywds, "s#|sO", keyword_argument_names, &s, &s_length, &encoding, &return_type)) {
+    return NULL;
+  }
+  /* NOTE(sigmavirus24): Because we use `s#` to parse the input string, we
```

Comment:
So the comment should say why `s#` and the answer is that `s` and `s*` are not consistent between python 2 and 3.